domain/src/main/java/org/oppia/android/domain/exploration/ExplorationProgressController.kt

```diff
@@ -735,4 +743,164 @@ class ExplorationProgressController @Inject constructor(
+        pendingStateHintList.forEachIndexed { index, hint ->
+          if (index < helpIndex.hintIndex) {
+            // Mark all hints as visible and revealed which have an index less than that stored in
+            // the HintState.
+            updatedHintList.add(hint.toBuilder().setHintIsRevealed(true).build())
+          } else if (index == helpIndex.hintIndex) {
+            // Add the currently shown hint to the updated hint list and set hintIsRevealed
+            // depending upon if the current hint was revealed or not.
+            updatedHintList.add(
+              hint.toBuilder().setHintIsRevealed(helpIndex.isAllVisibleHelpRevealed).build()
+            )
+          } else {
+            // Add all the remaining hints that are not yet visible to the user to the updated hint
+            // list.
+            updatedHintList.add(
+              hint.toBuilder().setHintIsRevealed(false).build()
+            )
+          }
+        }
+      }
+      HelpIndex.IndexTypeCase.SHOW_SOLUTION, HelpIndex.IndexTypeCase.EVERYTHING_REVEALED -> {
+        // All the hints are visible and revealed if helpIndex.indexTypeCase is equal to
+        // SHOW_SOLUTION or EVERYTHING_REVEALED.
+        pendingStateHintList.forEach { hint ->
+          updatedHintList.add(hint.toBuilder().setHintIsRevealed(true).build())
+        }
+      }
+      else -> updatedHintList.addAll(pendingStateHintList)
+    }
+    return updatedHintList
+  }
+
+  /**
+   * Set solution is reveled in the pendingState to true or false depending upon if solution was
+   * revealed for the current state pending state before the checkpoint was saved.
+   *
+   * @param pendingTopState the pending state created from the checkpoint
+   * @param helpIndex the state of solution for the exploration which was generated using the saved
+   *     checkpoint
+   */
+  private fun createSolutionFromCheckpoint(
+    pendingTopState: State,
+    helpIndex: HelpIndex
+  ): Solution {
+    return when (helpIndex.indexTypeCase) {
+      HelpIndex.IndexTypeCase.SHOW_SOLUTION -> {
+        pendingTopState.interaction.solution.toBuilder()
+          .setSolutionIsRevealed(false)
+          .build()
+      }
+      HelpIndex.IndexTypeCase.EVERYTHING_REVEALED -> {
+        pendingTopState.interaction.solution.toBuilder()
+          .setSolutionIsRevealed(true)
+          .build()
+      }
+      else -> pendingTopState.interaction.solution
+    }
+  }
+
+  /**
+   * Creates a list of completed states from the saved [ExplorationCheckpoint].
+   *
+   * @return [List] of [EphemeralState] containing all the states that were completed before the
+   *     checkpoint was created
+   */
+  private fun getPreviousStatesFromCheckpoint(
+    progress: ExplorationProgress
+  ): List<EphemeralState> {
+    val previousStates: MutableList<EphemeralState> = ArrayList()
+    progress.currentExplorationCheckpoint.completedStatesInCheckpointList
+      .forEachIndexed { index, state ->
```

Comment:
Per the above suggestion, you could use ``mapIndexed`` here, instead, to avoid the mutable list.

Reply:
changed to mapIndexed